### PR TITLE
Enc + TX + RX + Dec

### LINKED ISSUE #28: Merge `encode`/`decode` into `tx`/`rx` programs
Currently the full transmit/receive pipeline involves invoking 4 different programs. This is fairly cumbersome so it has been suggested to combine the encode/decoding process into the `tx`/`rx` programs. This way, `tx`/`rx` seamlessly handles the FEC encoding and decoding of data behind the scenes. 

Implementation strategy will be something like this: 

- Move the`tx.c:transmit_files` the loop should open the file, fec encode it into a buffer and then use the transmitters `transmit_bytes` function to transmit it. 
- `receiver.c` and `rx.c` will need to be updated to decode the data after capture. This will involve either using a temporary file that gets the capture data, and then is decoded. Or writing all the capture data into memory and then decoded all at once.

### PR DESCRIPTION
**Pull request by ghost**
Minor Memory leak due to not clearing `of_ldpc_staircase_set_fec_parameters`, however, want to push the major changes before figuring that one out.  Closes issue #28.

### CODE REVIEW

File: dxwifi/rx/rx.c [hunk 1]
```diff
@@ -146,17 +150,59 @@ dxwifi_rx_state_t setup_handlers_and_capture(dxwifi_receiver* rx, int fd) {
...
+        if(state != DXWIFI_RX_ERROR) {
+            //Open final file
+            //Error if file failed to open
+            if((fd = open(path, open_flags, mode)) < 0) {
+                log_error("Failed to open file: %s", path);
+            }
+            //Otherwise run FEC decoding.
+            else {
+                void *decoded_message = NULL;
+                size_t decoded_size = dxwifi_decode(encoded_data, temp_file_size, &decoded_message);
```
Comment (digital-nomad8620) [inline dxwifi/rx/rx.c:185]: We should check that `dxwifi_decode` was successful here and log the returned error if any. Pretty much exactly how `decode.c` does it: 

```c
    // Decode file
    void* decoded_msg = NULL;
    ssize_t msglen = dxwifi_decode(file_data, file_size, &decoded_msg);

    if(msglen > 0) {

        log_info("Successfully decoded %s. Decoded file size: %d", args->file_in, msglen);

        int nbytes = write(fd_out, decoded_msg, msglen);
        assert_M(nbytes == msglen, "Partial write occured: %d/%d - %s", nbytes, msglen, strerror(errno));
        free(decoded_msg);
    }
    else {
        log_error("Decode failed - %s", dxwifi_fec_error_to_str(msglen));
    
```

File: dxwifi/rx/rx.c [hunk 2]
```diff
@@ -146,17 +150,59 @@ dxwifi_rx_state_t setup_handlers_and_capture(dxwifi_receiver* rx, int fd) {
...
     mode_t mode     = S_IRUSR  | S_IWUSR | S_IROTH | S_IWOTH; 
+    
+    log_debug("Open File and Capture");
 
     dxwifi_rx_state_t state = DXWIFI_RX_ERROR;
-    if((fd = open(path, open_flags, mode)) < 0) {
-        log_error("Failed to open file: %s", path);
+    //Open temp file, with RW + Create
+    //Error if unable to open
+    if((temp_fd = open("tempfile", O_RDWR | O_CREAT , mode)) < 0) {
```
Comment (digital-nomad8620) [inline dxwifi/rx/rx.c]: Can we put the tempfile in the `/tmp` directory? The reason being is that `/tmp` is cleared on reboots and if something goes wrong on the satellite we don't want to accidently accrue a bunch of temporary files in the working directory. Or maybe just expose the temp file location as a command line argument so future maintainers can tweak it as necessary.

File: dxwifi/tx/cli.c [hunk 3]
```diff
@@ -212,6 +212,14 @@ static error_t parse_opt(int key, char* arg, struct argp_state *state) {
         args->tx.enable_pa = true;
         break;
 
+    case 'c':
+        args->coderate = atof(arg);
+        //Bounds Check
+        if(args->coderate > 1 || args->coderate < 0){
+            argp_error(state, "Error: Coderate must be a decimal between 0 and 1.");
+            argp_usage(state);
+        }
```
Comment (digital-nomad8620) [inline dxwifi/tx/cli.c:221]: You're missing a `break;` here.

File: test/test_tx_rx.py [hunk 4]
```diff
@@ -26,6 +26,8 @@
 from time import sleep
 from test.genbytes import genbytes
 
+# Constant: Size of ESI Symbol Size, in bytes.
+FEC_ESI = 1224
```
Comment (digital-nomad8620) [inline test/test_tx_rx.py]: ESI stands for 'Encoding Symbol ID' which is kind of a misnomer for this constant. As the comment says, this variable is the FEC Symbol size. Why not use FEC_SYMBOL_SIZE for the name of the constant?

File: test/test_tx_rx.py [hunk 5]
```diff
@@ -64,15 +66,15 @@ def tearDown(self):
         '''Remove files created during previous test'''
         shutil.rmtree(TEMP_DIR)
 
-
+    """
```
Comment (digital-nomad8620) [inline test/test_tx_rx.py]: Is there a reason this test case is commented out? We definitely shouldn't be commenting out test cases if they're failing. At the least, leave a very detailed comment describing why this test case is commented out and when we plan on fixing it.

File: test/test_tx_rx.py [hunk 6]
```diff
@@ -109,13 +111,13 @@ def testSingleFileTransmission(self):
         tx_out      = f'{TEMP_DIR}/tx.raw'
         rx_out      = f'{TEMP_DIR}/rx.raw'
 
-        tx_command = f'{TX} {test_file} -q --savefile {tx_out}'
-        rx_command = f'{RX} {rx_out} -q -t 2 --savefile {tx_out}'
+        tx_command = f'{TX} {test_file} -v --savefile {tx_out}'
```
Comment (digital-nomad8620) [inline test/test_tx_rx.py]: Can you turn off the verbosity for all these test cases? Having high verbosity is great for debugging locally, but makes scanning through the logs on the CI platform a pain.

File: test/test_tx_rx.py (hunk 4 above)
Comment (digital-nomad8620) [reply test/test_tx_rx.py]: Also, you need to pull in the latest changes from master. I believe you reviewed #39, in the description and code it was stated that we had to change the `RSCODE_NPAR` value, which effects the FEC symbol size. It changes it to `1099` I believe. I updated the [FEC Framing](https://app.diagrams.net/#G1OS1jDQYomK5IMrGjk3r-uisx0R9NhZec) document which should have all the latest values for FEC stuff.

**Review by digital-nomad8620 — CHANGES_REQUESTED** [review]
Requesting changes, please pull in the latest changes from Master which fixes the memory leak and changes/fixes the FEC stuff and verify all tests pass. Thanks!

**Comment by digital-nomad8620** [comment]
@CD-W Okay, everything looks good now. For future reference it's important to only call `munmap` on memory that was explicitly mapped with `mmap`. Otherwise it's just a death spiral of impossible bugs lol. Seriously, that was some of the weirdest memory bugs I've ever encountered. Once @khanu263  and @Cam57 get some on eyes on this and make sure everything's good we can merge.

**Review by digital-nomad8620 — APPROVED** [review]
Fixes approved

**Review by khanu263 — APPROVED** [review]
I did some manual runs and it seems to all work! No glaring issues that I can see in the code either. I'll make an issue to update the sweep script to only use the updated `tx` / `rx` binaries instead of `encode` / `decode`.

**Review by Cam57 — APPROVED** [review]
Looks great!

**Pull request merged**